lib/eco/nodeeditor.py

```diff
@@ -987,32 +987,31 @@ def createSubgrammarMenu(self, menu, change=False):
             # undo cursor change
             self.tm.cursor.node = tmp
 
-        settings = QSettings("softdev", "Eco")
-        if settings.value("gen_show_all_submenu").toBool() is False:
-            # Only show languages in submenu that are part of this composition,
-            # hiding all other languages known to Eco
-            todo = [self.get_mainlanguage()]
-            langs = []
-            while todo:
-                l = todo.pop(0)
-                for sl in lang_dict[l].included_langs:
-                    if lang_dict[sl] not in langs:
-                        langs.append(lang_dict[sl])
-                        todo.append(sl)
-        else:
-            langs = languages
+        # Filter out languages that are part of this composition.
+        todo = [self.get_mainlanguage()]
+        comp_langs = []
+        while todo:
+            l = todo.pop(0)
```

Comment:
Do you need to pop the `0`th element? That will cause quite a bit of shuffling vs. popping the last element.

Reply:
Not necessarily. I did this mainly to preserve the order in which the languages are composed into the main language. Though I guess it doesn't matter a great deal in which order they show up. On the other hand, this list is so small that this tiny bit of shuffling is not going to impact performance in any way.

Reply:
What order are they displayed in? Alphabetical?

Reply:
The order is the same as how the user composed them. So:
1. Main language
2. Sub lang 1
3. Sub lang 2
4. Sub lang 1 -> Sub lang
5. Sub lang 2 -> Sub lang

If we change the code to `todo.pop()`, then the order would be slightly different (reversing the sub-sub languages):
1. Main language
2. Sub lang 1
3. Sub lang 2
4. Sub lang 2 -> Sub lang
5. Sub lang 1 -> Sub lang

Admittedly, this is not a big deal.

Reply:
I wonder if we should do it alphabetically?

Reply:
I think the more immediate compositions should come first, though the order of any language coming after that doesn't really matter. So maybe we just use `todo.pop()`, which still shows the immediate sub languages first.

Reply:
What does "more immediate" mean?

Reply:
I wasn't sure how to describe this and the first thing that came to mind was "immediate". What I meant was the first level of the composition. So let's say I compose Html + Python + Sql, where Python is composed into HTML, and SQL compose into Python. So in this case I'd like Python to show up before HTML when opening the submenu, as it is comes earlier in "the order of the compositions", i.e. it is less nested, if that makes sense.

Reply:
If you're in HTML and you can only put Python into it, wouldn't SQL be put in the "other" menu?

Reply:
No, currently the submenu shows all languages that a part of this composition, no matter what language you are in. "Other" shows all other languages that Eco knows about that have nothing todo with the current composition.

Reply:
s/wouldn't/shouldn't/ ;)

Reply:
I still don't think it should be put there. The "other" menu should only contain languages that have nothing to do with the current composition (and thus should never be used here - which is why I excluded them all together in my initial commit). The main submenu contains all languages that are part of the composition, even if they can't be used in the current language box. This allows us to see at a quick glance, which languages we composed, and since languages valid at the current location are highlighted and move to the top, it don't think this is too confusing.

So basically there are 4 categories:
1. Languages valid at cursor position
2. Languages valid in current box
3. Languages that are part of this composition
4. All other languages Eco knows about

This PR currently merges 2. and 3. together, which IMO is ok.

Reply:
OK, I can live with 2+3 being merged together (presumably there's some sort of spacer between them and things in category 1?). But I think within each category, it's clearest if we present things alphabetically.

Reply:
Yes, there is a space between 1, 2/3, and 4. I will add a commit sorting the individual categories alphabetically.

Reply:
Thanks!

Reply:
The latest commit should do the trick.